geode-redis/src/distributedTest/java/org/apache/geode/redis/RedisSessionDistDUnitTest.java

```diff
@@ -0,0 +1,395 @@
+
+    cluster.startServerVM(indexOfServer2, redisPropsForServer2, locator.getPort());
+  }
+
+  @Test
+  public void should_getSessionDataFromServer_whenServerGoesDownAndIsRestarted() {
+    HttpEntity<String> request = new HttpEntity<>("noteFromClient2");
+
+    HttpHeaders resultHeaders = restTemplate
+        .postForEntity(
+            LOCALHOST + ":" + client2Port + "/addSessionNote",
+            request,
+            String.class)
+        .getHeaders();
+
+    String cookieString = resultHeaders.get("Set-Cookie").get(0);
+    HttpHeaders requestHeaders = new HttpHeaders();
+    requestHeaders.add("Cookie", cookieString);
+    HttpEntity<String> request2 = new HttpEntity<>("noteFromClient1", requestHeaders);
+
+    cluster.crashVM(indexOfServer2);
+
+    restTemplate
+        .postForEntity(
+            LOCALHOST + ":" + client1Port + "/addSessionNote",
+            request2,
+            String.class);
+
+    cluster.startServerVM(indexOfServer2,
+        redisPropsForServer2,
+        locator.getPort());
+
+    HttpEntity<String> request3 = new HttpEntity<>("", requestHeaders);
+
+    String[] sessionNotes = restTemplate
+        .exchange(
+            LOCALHOST + ":" + client2Port + "/getSessionNotes",
+            HttpMethod.GET,
+            request3,
+            String[].class)
+        .getBody();
+
+    assertThat(sessionNotes).containsExactly("noteFromClient2", "noteFromClient1");
+  }
+
+  @Test
+  public void should_getCorrectSessionData_whenGeodeRedisServerGoesDown_andClientConnectsToDifferentServer() {
+    HttpEntity<String> request = new HttpEntity<>("noteFromClient2");
+
+    HttpHeaders resultHeaders = restTemplate
+        .postForEntity(
+            LOCALHOST + ":" + client2Port + "/addSessionNote",
+            request,
+            String.class)
+        .getHeaders();
+
+    String cookieString = resultHeaders.get("Set-Cookie").get(0);
+    HttpHeaders requestHeaders = new HttpHeaders();
+    requestHeaders.add("Cookie", cookieString);
+    HttpEntity<String> request2 = new HttpEntity<>("", requestHeaders);
+
+    cluster.crashVM(indexOfServer2);
+
+    String[] sessionNotes = restTemplate
+        .exchange(
+            LOCALHOST + ":" + client2Port + "/getSessionNotes",
+            HttpMethod.GET,
+            request2,
+            String[].class)
+        .getBody();
+
+    assertThat(sessionNotes[0]).isEqualTo("noteFromClient2");
+
+    cluster.startServerVM(indexOfServer2, redisPropsForServer2, locator.getPort());
+  }
+
+
+  @Test
+  public void should_getCorrectSessionData_whenAppInstanceGoesDown_andClientConnectsToDifferentAppInstance() {
+
+    HttpEntity<String> request = new HttpEntity<>("noteFromClient2");
+    HttpHeaders responseHeaders = restTemplate
+        .postForEntity(
+            LOCALHOST + ":" + client2Port + "/addSessionNote",
+            request,
+            String.class)
+        .getHeaders();
+
+    String sessionAsCookie = responseHeaders.get("Set-Cookie").get(0);
```

Comment:
This block (http-level stuff) should be moved closer to where it's used, immediately above the try/finally block. (If http operations were all bundled into a method, then it wouldn't be possible to interleave logic like this.)